src/dataflow/src/render/delta_join.rs

```diff
@@ -341,146 +382,224 @@ where
-                *c = source_columns
-                    .iter()
-                    .position(|x| x == c)
-                    .expect("Column not found in source_columns");
-            }
-        })
-    }
-    for exprs in ready_equivalences.iter_mut() {
-        for expr in exprs.iter_mut() {
-            expr.visit_mut(&mut |e| {
-                if let ScalarExpr::Column(c) = e {
-                    *c = source_columns
-                        .iter()
-                        .position(|x| x == c)
-                        .expect("Column not found in source_columns");
+    /// Construct an instance of the closue from available columns.
+    ///
+    /// This method updates the available columns, equivalences, and
+    /// the `MapFilterProject` instance. The columns are updated to
+    /// include reference to any columns added by the application of
+    /// this logic, which might result from partial application of
+    /// the `MapFilterProject` instance.
+    ///
+    /// If all columns are available for `mfp`, this method works
+    /// extra hard to ensure that the closure contains all the work,
+    /// and `mfp` is left as an identity transform (which can then
+    /// be ignored).
+    pub fn build(
+        columns: &mut HashMap<usize, usize>,
+        equivalences: &mut Vec<Vec<ScalarExpr>>,
+        mfp: &mut MapFilterProject,
+    ) -> Self {
+        // First, determine which columns should be compare due to `equivalences`.
+        let mut ready_equivalences = Vec::new();
+        for equivalence in equivalences.iter_mut() {
+            if let Some(pos) = equivalence
+                .iter()
+                .position(|e| e.support().into_iter().all(|c| columns.contains_key(&c)))
+            {
+                let mut should_equate = Vec::new();
+                let mut cursor = pos + 1;
+                while cursor < equivalence.len() {
+                    if equivalence[cursor]
+                        .support()
+                        .into_iter()
+                        .all(|c| columns.contains_key(&c))
+                    {
+                        // Remove expression and equate with the first bound expression.
+                        should_equate.push(equivalence.remove(cursor));
+                    } else {
+                        cursor += 1;
+                    }
                 }
-            });
-        }
-    }
-
-    // Apply a filter if either list of constraints is non-empty.
-    if ready_to_go.is_empty() && ready_equivalences.is_empty() {
-        (updates, None)
-    } else {
-        let (ok_collection, err_collection) = updates.filter_fallible(move |input_row| {
-            let temp_storage = repr::RowArena::new();
-            let datums = input_row.unpack();
-            for p in &ready_to_go {
-                if p.eval(&datums, &temp_storage)? != Datum::True {
-                    return Ok(false);
+                if !should_equate.is_empty() {
```

Comment:
should this check for `if should_equate.len() > 1` instead? if not I think a comment explaining why would be nice as the code in `apply` looks like a no-op for single element equivalence classes.

I had this thought as I was falling asleep so it may be no good :)